Web tutorial for Msig Txn (#847)
scale-it/algo-builder: docs/tutorials/t-09.md

## Steps

In this tutorial we will use Algo Builder Wallet Widget to connect to various web wallets. Then we will show how to instrument the wallet to sign a multisignature transaction.
See [Wallet Widget tutorial](https://developer.algorand.org/solutions/vuejs-multi-wallet-widget/) for more details about the widget.

### MultiSignature Transaction

To create a msig transaction, you can follow steps from the offical documentation: [Create Multisignature](https://developer.algorand.org/docs/get-details/accounts/create/#multisignature) and [Transaction](https://developer.algorand.org/docs/sdks/javascript/) to use it.
Here is a glimpse of creating the multisig transaction:

```js
const mparams = {
	version: 1,
	threshold: 2,
	addrs: [account1.addr, account2.addr, account3.addr],
};

let multsigaddr = algosdk.multisigAddress(mparams);
let msigTxn = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
	from: multsigaddr,
	to: receiver,
	amount: amount,
	note: note,
	suggestedParams: params,
});
```

After creating the transaction, you can sign it using different wallets.

> 1. AlgoSigner<br/>
>    We use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing `AlgoSigner` and the Chain Name i.e the network name. The `web` instance will allow us to sign and send transactions to the blockchain.

Reviewer: ```suggestion
#### AlgoSigner

We use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing `AlgoSigner` and the Chain Name i.e the network name. The `web` instance will allow us to sign and send transactions to the blockchain.
```

```js
const web = new WebMode(AlgoSigner, CHAIN_NAME);
```

WebMode method `signMsigTxn` expects txn in [`EncodedSignedTransaction`](https://algorand.github.io/js-algorand-sdk/interfaces/EncodedSignedTransaction.html) type. To convert the above created msig transaction, below code is used:

```js
let subsig = [];
for (const address of mparams.addrs) {
	subsig.push({ pk: decodeAddress(address).publicKey });
}
const msigEncodedTxn = {
	msig: { v: mparams.version, thr: mparams.threshold, subsig: subsig },
	txn: msigTxn.get_obj_for_encoding(),
};
```

To sign the msig transaction:

```js

> const signedTxnResponse = await web.signMsigTxn([msigEncodedTxn]);

Reviewer: where `signMsigTxn` is defined? I don't see that function in algo-builder

Reviewer: And why it takes a list?

Author: https://github.com/scale-it/algo-builder/blob/develop/packages/web/src/lib/web-mode.ts#L192

Reviewer: 1. This function is called `signMsigTransaction`. I've made a PR to today to rename it to `signMsigTxn` (as it's usded here)
2. That function doesn't take a list as the first argument.

Author: oh my bad, I added it as `signMsigTxn  but during review Stan asked to update it to `signMsigTransaction` and I forgot to update it here

> ```

Reviewer: is this going to sign with all available accounts, or only with the account currently seclected in algosigner? What if the selected account is not part of the multisig? Let's clarify this in the description here.

> The following screen will popup for you to sign the txn:

Reviewer 2: ```suggestion
The below screen should popup for you to sign the txn:
```

Author: what's the issue with using following word?

<br/>
<img src="./t-09/assets/algosigner.png" height="500" title="Sign Msig Txn" />

> After signing, the response is the object containing blob key, which we can use further to send the signed transaction to the network.

Reviewer: ```suggestion
After signing, the response is the object containing signed transaction data in a `blob` field, which we can use to submit to the network.
```

To send the above txn we need to first convert it to base64:

```js
const base64Txn = Buffer.from(signedTxnResponse.blob).toString("base64");
const sendTxnResponse = await webmode.sendAndWait(base64Txn);
```

The sent txn response includes the confirmed transaction response along with txID.

> 2. MyAlgo Wallet <br/>
>    We use `connectMyAlgoWallet()` method to connect to My Algo wallet.

Reviewer: ```suggestion
#### MyAlgo Wallet

We use `connectMyAlgoWallet()` method to connect to My Algo wallet.
```

```js
   async connectMyAlgoWallet() {
      try {
        let walletURL = {
          token: "",
          server: "https://node.algoexplorerapi.io",
          port: "",
        }
        let myAlgoMode = new MyAlgoWalletSession(walletURL);
        await myAlgoMode.connectToMyAlgo();
        if (myAlgoMode.accounts.length) {
          this.walletAddress = myAlgoMode.accounts[0].address;
        }
      } catch (e) {
        console.error(e);
      }
   }
```

Algo Wallet method `signTransaction` expects transaction in [algosdk.Transaction](https://algorand.github.io/js-algorand-sdk/classes/Transaction.html) type.

```js

> const txn = algosdk.Transaction.from_obj_for_encoding(msigEncodedTxn.txn);

Reviewer: I think this equals to `let msigTxn = algosdk.makePaymentTxnWithSuggestedParamsFromObject`. So, instead of recreating it, let's reuse `msigTxn` here.

const signedTxnResponse = await myAlgoMode.signTransaction(txn);
```

> The following screens will popup for you to login to msig and sign the txn:<br/>

Reviewer 2: ```suggestion
The below screens should popup for you to login to msig and sign the txn:<br/>
```

<img align="left" src="./t-09/assets/myalgo-1.png" height="500" title="Unlock msig" />
<img align="left" src="./t-09/assets/myalgo-2.png" height="500" title="Sign Msig Txn" />
<img  src="./t-09/assets/myalgo-3.png" height="500" title="Pending confirmation" />

The signed txn response is the object containing blob key.

> To send the above txn:

Author: the function is already added to web package, we merge the multi signatures according to the input provided, it is all done in `signMsigTransaction` itself

```js
const sendTxnResponse = await algoWallet.sendAndWait(signedTxnResponse.blob);
```

Here also the sent txn response includes the confirmed transaction response along with txID.

> 3. Wallet Connect<br/>
>    Perra Wallet currently doesn't support Multisig transaction signature, thus it's support will be added after Perra wallet supports it.

Reviewer: ```suggestion
#### Wallet Connect

Perra Wallet currently doesn't support Multisig transaction signature, thus it's support will be added after Perra wallet supports it.
```

	suggestedParams: params,
});
```

After creating the transaction, you can sign it using different wallets.

#### AlgoSigner

We use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing `AlgoSigner` and the Chain Name i.e the network name. The `web` instance will allow us to sign and send transactions to the blockchain.

```js
const web = new WebMode(AlgoSigner, CHAIN_NAME);
```

WebMode method `signMsigTxn` expects txn in [`EncodedSignedTransaction`](https://algorand.github.io/js-algorand-sdk/interfaces/EncodedSignedTransaction.html) type. To convert the above created msig transaction, below code is used:

```js
let subsig = [];
for (const address of mparams.addrs) {
	subsig.push({ pk: decodeAddress(address).publicKey });
}
const msigEncodedTxn = {
	msig: { v: mparams.version, thr: mparams.threshold, subsig: subsig },
	txn: msigTxn.get_obj_for_encoding(),
};
```

To sign the msig transaction:

```js
const signedTxnResponse = await web.signMsigTxn([msigEncodedTxn]);
```

The following screen will popup for you to sign the txn:
<br/>
<img src="./t-09/assets/algosigner.png" height="500" title="Sign Msig Txn" />

> After signing, the response is the object containing signed transaction data in a `blob` field, which we can use to submit to the network.

Reviewer: ```suggestion
After signing, the returned object contains signed transaction data in a `blob` field, which we can use to submit to the network.
```

```js
const msigEncodedTxn = tx.mkMultisigEncodedTx(mparams, msigTxn);
```

To sign the msig transaction, make sure to login and sign from the account which is part of the msig transaction, else it will throw an error:

```js
const partialSignedTx = await web.signMsigTx(msigEncodedTxn);
```

The following screen will popup for you to sign the txn:
<br/>
<img src="./t-09/assets/algosigner.png" height="500" title="Sign Msig Txn" />

After signing, the response is the object containing blob key which in base64 format, which we can use further to send the signed transaction to other co-signers and also to send it to the network.

For the co-signer to sign the transaction he/she can decode the `partialSignedTx` and sign using:

```js
const decodedPartialSignedTx = algosdk.decodeSignedTransaction(
	Uint8Array.from(Buffer.from(partialSignedTx.blob, "base64"))
);
const signedTxnResponse = await web.signMsigTransaction(decodedPartialSignedTx);
```

> To send the signed txn you can use `sendAndWait` method:

Reviewer: ```suggestion
Once we have enough signatures (equal or more than the msig threshold),  anyone can send the final txn using `sendAndWait` method:
```

```js

> const sendTxnResponse = await webmode.sendAndWait(signedTxnResponse.blob);

Reviewer: ```suggestion
const sendTxnResponse = await webmode.sendAndWait(signedTx.blob);
```

   async connectMyAlgoWallet() {
      try {
        let walletURL = {
          token: "",
          server: "https://node.algoexplorerapi.io",
          port: "",
        }
        let myAlgoMode = new MyAlgoWalletSession(walletURL);
        await myAlgoMode.connectToMyAlgo();
        if (myAlgoMode.accounts.length) {
          this.walletAddress = myAlgoMode.accounts[0].address;
        }
      } catch (e) {
        console.error(e);
      }
   }
```

Algo Wallet method `signTransaction` expects transaction in [algosdk.Transaction](https://algorand.github.io/js-algorand-sdk/classes/Transaction.html) type.

```js
const partialSignedTx = await myAlgoMode.signTransaction(msigTxn);
```

The following screens will popup for you to login to msig and sign the txn:<br/>
<img align="left" src="./t-09/assets/myalgo-1.png" height="500" title="Unlock msig" />
<img align="left" src="./t-09/assets/myalgo-2.png" height="500" title="Sign Msig Txn" />
<img  src="./t-09/assets/myalgo-3.png" height="500" title="Pending confirmation" />

The signed txn response is the object containing blob key.

You can send the blob key which in Uint8Array format to co-signer to sign the transaction:

```js

> const decodedPartialSignedTx = algosdk.decodeSignedTransaction(partialSignedTx.blob, "base64");
> const signedTxnResponse = await myAlgoMode.signTransaction(decodedPartialSignedTx.txn);

Reviewer: ```suggestion
const decodedPartiallySignedTx = algosdk.decodeSignedTransaction(partialSignedTx.blob, "base64");
const signedTx = await myAlgoMode.signTransaction(decodedPartiallySignedTx.txn);
```

Reviewer: make similar renames above and below.
